### ISSUE TITLE
Seeking new maintainers for gl-rs

### ISSUE THREAD

**Issue by brendanzab** [brendanzab/gl-rs#524]
I'm currently don't have the mental bandwidth to maintain this library! Most of my current work is in programming languages, and it's a bit of a context switch to jump back to gl-rs, and so I end up procrastinating on looking at pull requests or issues! This is not a good position for such a critical library to be in, even given the current shift to newer APIs like Vulkan. I'm really sorry if this has affected you if you are reading this!

Going forward I'd like to see if I can find a better home for gl-rs. It was started a long time ago ([8 years ago](https://github.com/brendanzab/gl-rs/tree/b43b6ff55f8165ca177adfb65b724da4959180a9) in fact!), and could probably do with some TLC. Please let me know if you'd be up for helping out with this, or have any ideas for where to migrate this repository! The help would be most appreciated! A plan for crate stewardship was mentioned in https://github.com/rust-gamedev/wg/issues/42, but I haven't got around to following up until now.

**Comment by Lokathor** [comment]
The gamedev-wg is still interested in being a home for this repository, but others should have a chance to speak up with ideas before any big moves happen.

**Comment by brendanzab** [comment]
Thanks for the quick comment, yeah I thought it best I post here and get the discussion started.

**Comment by AlexEne** [comment]
Do you have some outstanding issues/plans that you'd think are a bit more high-priority to be done? It does seem like a light workload if we split it with the WG members, at only 17 issues (I didn't read them so maybe there's years of work in there :D )

**Comment by brendanzab** [comment]
Good question. I'm guessing something like:

- Review and merge outstanding PRs
- Switch to Github actions
- Check if the webgl binding generator is still needed (I think folks are using something else these days?), and if not deprecate it
- Update any dependencies, and ensure that we are still using the ones recommended in the community
- Review versioning strategy and repository structure - at the moment all crates are versioned independently in a monorepo, which is painful for a maintainer to manage

@Lokathor also attempted to do a cleanup in #507, but ran up against some frustrations with the test suite.

**Comment by AlexEne** [comment]
I'm fine with the gamedev-wg move if it's needed.

**Comment by MarijnS95** [comment]
@brendanzab is this offer still open? I won't be able to dedicate much maintenance effort but lots of high-value crates are still using the `gl` and `gl-generator` crates, including `glutin`. We could use a bindings update such as the one submitted in #546 for example, and could clear out some outstanding issues/PRs.

**Comment by ids1024** [comment]
Probably most users of gl/gl-generator are using glutin. It could make sense to maintain under the `rust-windowing` alongside glutin, although it isn't specifically windowing related.

Gfx-rs is also a relevant organization (that I'm not a part of) but wgpu seems to use gl-generator only indirectly by depending on glutin.

**Comment by MarijnS95** [comment]
The only other popular crates in my search for GL on crates.io are `glow` and `khronos-egl`, the latter also uses `gl` but hand-writes its EGL bindings instead of using the bindings generator here. Perhaps some of the workload could be shared there.

**Comment by Lokathor** [comment]
Last I was aware, `glow` does not use the `gl` crate, but instead bundles the output of my own generator crate `phosphorus`

**Comment by MarijnS95** [comment]
Fwiw I was referring to popular crates that provide GL bindings, not necessarily through crates provided by this repository. To make a case for rather maintaining what everyone is currently still using, before too many new forks/crates are spawned to slowly diverge the ecosystem.

`phosphorus` was mentioned to me some time ago, anything that can be done to bring the ecosystem forward as a whole here?

**Comment by ids1024** [comment]
Would `phosphorus` work for EGL/WGL/GLX like `gl_generator`? I guess if they're all just C functions and the same xml format?

If it can handle all of these, perhaps `glutin` could make use it.

**Comment by MarijnS95** [comment]
@ids1024 I'm quite sure it doesn't, looking at the files in the repository and what was mentioned in https://github.com/rust-windowing/glutin/issues/1653#issuecomment-1861420287. Hence my suggestion/request that there might be something to win by unifying.

**Comment by Lokathor** [comment]
It could probably be updated to work on more stuff.

Currently it's used literally just by me and groves (the `glow` lead), so it doesn't do anything else because no one asked. However, I suspect it's probably pretty trivial to add EGL support.

**Comment by ids1024** [comment]
Yeah, probably not too hard to add.

There's also https://github.com/Dav1dde/glad, which added Rust support. And seems to be a pretty popular GL generator for C at the moment. It's written in Python though.

**Comment by MarijnS95** [comment]
@Lokathor what would be the advantage of `phosphorus` over `gl-generator`? Assuming we can figure out maintainership (proposed by the author themselves in this specific issue) and update the bindings source, I'd rather stick to what is currently being used and known to work (perhaps with some incremental improvements) than switch to something completely new.

**Comment by MarijnS95** [comment]
Any updates on transitioning ownership to a larger organization with a bigger bus-factor?

**Comment by kchibisov** [comment]
As alternative you can move it to `rust-windowing`, which hosts glutin/raw-window-handle and has plenty of maintainers.

**Comment by MarijnS95** [comment]
Ping again @brendanzab, is there anything we can get going here?

**Comment by djc** [comment]
An unmaintained advisory was just submitted to the advisory-db (see https://github.com/rustsec/advisory-db/pull/2306). I've reached out to @brendanzab via email to ask how to they'd like to proceed (the plan to transition this to rust-windowing sounds good to me).

**Comment by brendanzab** [comment]
@MarijnS95 if you’re still open to it, I can add you as a collaborator, and as an owner on crates.io? Then we can consider where we move the repository to keep things ticking over. I’ve given a way a decent number of crates over the years so it shouldn’t be an issue.

**Comment by MarijnS95** [comment]
I'm sorry to hear that @brendanzab, wishing you all the best!

Not sure about activity on the `gl` crate, but the `gl_generator` crate (and relevant `khronos_api` dependency) is still in active use by `glutin` and I'd like to maintain those a bit going forward to improve usability in `glutin`.

Thanks for all the invites, I've accepted them and will start the transition process to `rust-windowing` after a quick check with the other maintainers. After that, we can look into merging some outstanding PRs and blowing some fresh life into this 👍

**Comment by brendanzab** [comment]
I think in the case of `gl` it would probably be fine to deprecate it and point people to better alternatives, if it’s no longer serving any use.

**Comment by MarijnS95** [comment]
Thanks, I'll take a look at `gl` and consider adding a deprecation notice whenever we're ready to publish updates of all crates.

For now it seems you've invited me as collaborator which doesn't give me access to https://github.com/brendanzab/gl-rs/settings, which I'd need to initiate the transfer.

---

@djc perhaps you meant to write `glow` - both it and `gl` provide bindings to GL functions (as does `gl_generator` directly within a project, such as `glutin`'s `-sys` crates). `glutin` instead provides abstractions over the platform-specific bindings (EGL, WGL, GLX, CGL...).

**Comment by djc** [comment]
> [@djc](https://github.com/djc) perhaps you meant to write `glow` - both it and `gl` provide bindings to GL functions (as does `gl_generator` directly within a project, such as `glutin`'s `-sys` crates). `glutin` instead provides abstractions over the platform-specific bindings (EGL, WGL, GLX, CGL...).

Sure -- I'm not actually familiar with the subject matter.

**Comment by brendanzab** [comment]
I can initiate a transfer to `rust-windowing` if we're ready to do so. Just let me know what I need to do :)

Seems like the GDK4 bindings use `gl`, including a few other crates: https://lib.rs/crates/gl/rev

**Comment by MarijnS95** [comment]
@brendanzab we received the go-ahead from @kchibisov in the Winit Matrix chat, feel free to initiate it!

`khronos-egl` only uses `gl` in `dev-dependencies`, that can be updated but at the same time that crate appears stale and provide the same functionality as `glutin` (without platform-agnostic bindings). `wgpu-hal` is the most significant user, which we can try to migrate to `glutin`. It relates to some old issues on their end that I was pulled in to, and they would benefit significantly from replacing their open-coded GL context creation/activation with `glutin` entirely: https://github.com/gfx-rs/wgpu/issues/5505 / https://github.com/gfx-rs/wgpu/issues/6211.

`gdk4` only has a very tiny `gl` binding for the `GLSync` type (used in `GLTextureBuilder`) which is otherwise defined to be a `*const c_void`: https://docs.rs/gdk4/latest/gdk4/struct.GLTextureBuilder.html

I still need to check the other crates for their significance of usage.

**Comment by djc** [comment]
> [@brendanzab](https://github.com/brendanzab) we received the go-ahead from [@kchibisov](https://github.com/kchibisov) in the Winit Matrix chat, feel free to initiate it!

AIUI, transferring a repo to an org generally needs the transferring user to have permissions in the org. That's why I suggested it might be easier to transfer to @MarijnS95 as an intermediate step.

**Comment by brendanzab** [comment]
Will that cause much churn for downstream people? 🤔

**Comment by djc** [comment]
> Would going via Marijin cause much churn for downstream people? 🤔

No, repo transfers generally don't cause much churn. GitHub puts redirects in place and it would probably just be a few minutes before @MarijnS95 moves it over.

**Comment by MarijnS95** [comment]
That's a good point. I've renamed my fork (which exists because of an outstanding PR contribution) to make this possible, but am afraid `gh pr create` might in some distant future create it anew if I ever decide to remove that fork entirely.

I wonder if GitHub coalesces the redirects if the repository is transferred twice.

**Comment by brendanzab** [comment]
Transfer complete!

Thanks everyone for your patience with me. It’s really cool to think this silly thing I started over a decade ago is still useful to people today, at least for the time being!
